subprojects/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/MavenConversionIntegrationTest.groovy

```diff
@@ -649,6 +649,56 @@ Root project 'webinar-parent'
         dsl.getBuildFile().text.contains(TextUtil.toPlatformLineSeparators(mavenLocalRepoBlock))
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/20981")
+    def "escapeSingleQuotes"() {
+        def dsl = dslFixtureFor(scriptDsl)
+
+        when:
+        run 'init', '--dsl', scriptDsl.id as String
+
+        then:
+        dsl.assertGradleFilesGenerated()
+
+        def isGroovy = scriptDsl == BuildInitDsl.GROOVY
+        def descriptionPropertyAssignment = (isGroovy ? 'description = \'That\\\'s it\'' : 'description = "That\'s it"')
+        dsl.getBuildFile().text.readLines().contains(descriptionPropertyAssignment)
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/20981")
+    def "escapeDoubleQuotes"() {
+        def dsl = dslFixtureFor(scriptDsl)
+
+        when:
+        run 'init', '--dsl', scriptDsl.id as String
+
+        then:
+        dsl.assertGradleFilesGenerated()
+
+        def isGroovy = scriptDsl == BuildInitDsl.GROOVY
+        def descriptionPropertyAssignment = (isGroovy ? 'description = \'"Quoted description"\'' : 'description = "\\"Quoted description\\""')
+        dsl.getBuildFile().text.readLines().contains(descriptionPropertyAssignment)
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/20981")
+    def "escapeBackslashes"() {
+        def dsl = dslFixtureFor(scriptDsl)
+
+        when:
+        run 'init', '--dsl', scriptDsl.id as String
+
+        then:
+        dsl.assertGradleFilesGenerated()
+
+        def isGroovy = scriptDsl == BuildInitDsl.GROOVY
+        '''hello world ' this is cool!'''
+        def descriptionPropertyAssignment = (
+            isGroovy
+                ? 'description = \'Backslashes \\\\ next to single quotes \\\'\\\\\\\' and double quotes "\\\\" oh my!\''
+                : 'description = "Backslashes \\\\ next to single quotes \'\\\\\' and double quotes \\"\\\\\\" oh my!"'
```

Comment:
I find it difficult to read these assertions. A couple of other possibilities crossed my mind:

First, using Groovy triple-quoted strings. This way, we wouldn't have to escape most of the single quotes:

```suggestion
                ? '''description = 'Backslashes \\\\ next to single quotes \\'\\\\\\' and double quotes "\\\\" oh my!\''''
                : 'description = "Backslashes \\\\ next to single quotes '\\\\' and double quotes \\"\\\\\\" oh my!"'
```

Second, using some kind of templating approach:

```suggestion
                ? 'description = \'Backslashes $BACKSLASH next to single quotes $SINGLE_QUOTE$BACKSLASH$SINGLE_QUOTE and double quotes "$BACKSLASH" oh my!\''
                : 'description = "Backslashes $BACKSLASH next to single quotes \'$BACKSLASH\' and double quotes \\"$BACKSLASH\\" oh my!"'
```

(I know this isn't valid Groovy.)

Finally, maybe reading the expected strings from a file, where we wouldn't have to use escape characters:

```
description = 'Backslashes \\ next to single quotes \'\\\' and double quotes "\\" oh my!'
```
```
description = "Backslashes \\ next to single quotes '\\' and double quotes \"\\\" oh my!"
```

Reply:
One simple thing would be to check the backslash and single/double quote in separate test cases. 

I like the idea of using triple quotes to reduce escaping. 👍

Reply:
Re: checking the backslash and quotes in separate test cases, I wanted to add a test case that used both backslashes and quotes together. That way, the test checks that we're replacing `\` with `\\`, then replacing `"` with `\"`. If we replaced `"` before replacing `\`, that could cause bugs. E.g. the escaping logic should transform `\"` into `\\\"`. However, if it did the replacements in the wrong order, it'd transform `\"` into `\\\\"` instead.

Maybe this is better suited for a unit test. We can have this integration test check a simple case. 

And sounds good, I'll change this test to use triple quotes.